distributed/executor.py

```diff
@@ -976,6 +987,66 @@ def cancel(self, futures):
         return sync(self.loop, self._cancel, futures)
 
     @gen.coroutine
+    def _publish_dataset(self, data, name):
+        keys = [tokey(f.key) for f in futures_of(data)]
+        yield self.scheduler.publish_dataset(keys=keys, name=tokey(name),
+                                             data=pickle.dumps(data),
+                                             client=self.id)
+
+
+    def publish_dataset(self, data, name):
+        """
+        Store a named reference to the data in the scheduler for other executors
+
+        Parameters
+        ----------
+        data: list of futures or dask collection
+        name: str or container
+            handle by which the stored data should be known. If not a string,
+            will be wrapped by tokey() to deterministically turn into one.
+
+        Returns
+        -------
+        None
+        """
+        sync(self.loop, self._publish_dataset, data, name)
+
+    @gen.coroutine
+    def _unpublish_dataset(self, name):
+        yield self.scheduler.unpublish_dataset(name=name)
+
+    def unpublish_dataset(self, name):
+        """
+        Remove reference to data on the scheduler
+        """
+        sync(self.loop, self._unpublish_dataset, name)
+
+    @gen.coroutine
+    def _published_datasets(self):
+        resp = yield self.scheduler.get_published_keys()
+        raise Return(resp)
+
+    def published_datasets(self):
+        """ Returns list of named datasets referenced in the scheduler
+
+        See publish_dataset()
+        """
+        return sync(self.loop, self._published_datasets)
+
+    @gen.coroutine
+    def _get_published_dataset(self, name):
+        out = yield self.scheduler.get_published_dataset(name=name,
+                                                         client=self.id)
+        data, keys = out['data'], out['keys']
+
+        with temp_default_executor(self):
+            data = pickle.loads(data)
+        raise Return(data)
+
+    def get_published_dataset(self, name):
+        return sync(self.loop, self._get_published_dataset, tokey(name))
```

Comment:
Ah, after looking below I see the difference.  I recommend the following names:

*  get_published_keys -> list_datasets
*  get_published_dataset -> get_dataset
*  publish_dataset

The term "key" generally means the name of a task.  It would be nice to not overload this term.  I suggest that we now marry the term "dataset" for all functions related to what you're doing now.  The term "list" might be a nice alternative to "get", when we're just getting the names of available datasets.